Mobile filter

dirkhh opened the pull request:
<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->
This is mainly code originally done by @janmulder  and then massaged into place by me.
It adds filtering to the mobile UI.
So far only tested on mobile-on-desktop. I'll look at it on device between dives later, but I wanted to get it out for the Europeans to play with.

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->
1) add a UI action to toggle the filter (this is all Jan's work)
2) add a backend to apply the filtering (mostly Jan's)
3) changed that backend to do full text search instead of just dive sites
4) added a count of filtered dives to the header (and made it an overlay)

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->
The count on top should be correct, but the count that is part of the trip headers is wrong - it shows the count of all dives (filtered or not) in the trip. That's a bit confusing and needs to be fixed.

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->
TODO

### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->
TODO

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->
@janmulder @jbygdell and anyone else who plays with the mobile UI
Which soon will be @mturkia :-)

dirkhh commented:
1. as I said, only tested on laptop where performance and focus weren't an issue. Adding an apply button would be very easy
2. also doable, simply requires a transition for the visual effect; what's harder is to have it push the list down and not cover the first dive
3. this turns into a user preference / number of options thing; I personally would want the search to be case insensitive. If I search for "frog fish" I don't care if it's at the beginning of a sentence or not. I also know that people will start arguing about searching the beginning of certain text strings only (we had that argument about location names before). Etc. I'm happy to entertain certain options here, but it has to stay reasonable. Especially if we add an 'apply' button we will already start getting tight on space - and the last thing I want to do is turn this into two lines.

mturkia commented:
Isn't the full text missing notes?

On Wed, Oct 17, 2018, 13:52 Dirk Hohndel <notifications@github.com> wrote:

> Describe the pull request:
>
>    - Bug fix
>    - Functional change
>    - New feature
>    - Code cleanup
>    - Build system change
>    - Documentation change
>    - Language translation
>
> Pull request long description:
>
> This is mainly code originally done by @janmulder
> <https://github.com/janmulder> and then massaged into place by me.
> It adds filtering to the mobile UI.
> So far only tested on mobile-on-desktop. I'll look at it on device between
> dives later, but I wanted to get it out for the Europeans to play with.
> Changes made:
>
>    1. add a UI action to toggle the filter (this is all Jan's work)
>    2. add a backend to apply the filtering (mostly Jan's)
>    3. changed that backend to do full text search instead of just dive
>    sites
>    4. added a count of filtered dives to the header (and made it an
>    overlay)
>
> Related issues: Additional information:
>
> The count on top should be correct, but the count that is part of the trip
> headers is wrong - it shows the count of all dives (filtered or not) in the
> trip. That's a bit confusing and needs to be fixed.
> Release note:
>
> TODO
> Documentation change:
>
> TODO
> Mentions:
>
> @janmulder <https://github.com/janmulder> @jbygdell
> <https://github.com/jbygdell> and anyone else who plays with the mobile UI
> Which soon will be @mturkia <https://github.com/mturkia> :-)
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Subsurface-divelog/subsurface/pull/1808
> Commit Summary
>
>    - Mobile/filtering: add icons for filter and sort
>    - Mobile/filtering: first attempt to filter on dive site
>    - Mobile/filtering: full text filter, instead of just dive site
>    - Mobile/filtering: make the search box an overlay
>    - Mobile/filtering: add count of filtered dives to search bar
>
> File Changes
>
>    - *M* core/subsurface-qt/DiveObjectHelper.cpp
>    <https://github.com/Subsurface-divelog/subsurface/pull/1808/files#diff-0>
>    (6)
>    - *M* core/subsurface-qt/DiveObjectHelper.h
>    <https://github.com/Subsurface-divelog/subsurface/pull/1808/files#diff-1>
>    (2)
>    - *M* mobile-widgets/qml/DiveList.qml
>    <https://github.com/Subsurface-divelog/subsurface/pull/1808/files#diff-2>
>    (56)
>    - *A* mobile-widgets/qml/icons/ic_filter_list_24px.svg
>    <https://github.com/Subsurface-divelog/subsurface/pull/1808/files#diff-3>
>    (1)
>    - *A* mobile-widgets/qml/icons/ic_sort_24px.svg
>    <https://github.com/Subsurface-divelog/subsurface/pull/1808/files#diff-4>
>    (1)
>    - *M* mobile-widgets/qml/main.qml
>    <https://github.com/Subsurface-divelog/subsurface/pull/1808/files#diff-5>
>    (3)
>    - *M* mobile-widgets/qml/mobile-resources.qrc
>    <https://github.com/Subsurface-divelog/subsurface/pull/1808/files#diff-6>
>    (2)
>    - *M* qt-models/divelistmodel.cpp
>    <https://github.com/Subsurface-divelog/subsurface/pull/1808/files#diff-7>
>    (19)
>    - *M* qt-models/divelistmodel.h
>    <https://github.com/Subsurface-divelog/subsurface/pull/1808/files#diff-8>
>    (6)
>
> Patch Links:
>
>    - https://github.com/Subsurface-divelog/subsurface/pull/1808.patch
>    - https://github.com/Subsurface-divelog/subsurface/pull/1808.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Subsurface-divelog/subsurface/pull/1808>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ACjPRAWyD-PIAsK6wbfcr5ahwRtYM61Tks5ulwvrgaJpZM4XjpWX>
> .
>

janmulder commented:
1) I have focus issues on desktop as well. Not tried it on device yet.
2) Like I said: pure UI, not the most important thing at this point.
3) Agree. Not a multi line thing.

new point 4) set filter on. Search, select a dive far away (old), set filter off. As the dive is far out of the view presented, it starts scrolling. Which takes forever on a big logbook.

new point 5) see 4), but now we arrived at the chosen dive. The trip where that dive is in is collapsed, so you do not see the dive you have selected before. Very confusing.

janmulder commented:
@mturkia: currently it does not look in notes. Not sure this is by design or just forgotten.

bstoeger commented:
Not directly related to this PR - but can someone explain why we need a separate list-model for mobile?

dirkhh commented:
- clear button - let's reset the filter when toggling and see how that works.
- focus issues - I may need help with that, let's see
- I'm sure how you get 52 results for "tes" on mobile but not in Subsurface - but I have a suspicion. I'll play with that.
Thanks to both of you for the testing. Really appreciated

dirkhh commented:
As always with QML, things that should be easy take an inordinately long time while you try to figure out why it doesn't do what you want it to do.
Here's what I have implemented so far:
- filter animation: this should work as it exactly matches other animations we use, but at least when run on the desktop it doesn't look the way I expected it to look. It also still doesn't push down the list (so it covers the first item)... now that I write this I have a brilliant idea... why is this a Header and not a Footer... maybe that would look better as well. Oh, wait, but that then collides with the action button, doesn't it? darn.
- filter clears when toggled
- filtering only starts when you hit enter
- I added notes to the full text search and also added a separator between the concatenated strings. My suspicion is that when @jbygdell got weird match counts it was matching the end of one string plus the beginning of the next string. E.g., do you have buddies with last letter 't' and dive masters starting with 'es' (or 'te' and 's')?

As usual, I'd love to get this in more people's hands earlier, so let's please focus on what has to be fixed before this can go into master / create beta test binaries... (but let's still keep track of everything else that needs to be addressed... especially the scrolling issue worries me)

dirkhh commented:
@bstoeger 
> Not directly related to this PR - but can someone explain why we need a separate list-model for mobile?

I'm guessing this was because of the complexity of the list/tree model we use on the desktop

jbygdell commented:
I spent some time last night trying to fix the focus issues, one is still not fixed.
This is the behavior I'm seeing on my Android device:
When the filter view becomes visible, it is not active and the keyboard is not visible.
So now I have to tap in the text input for this to happen.

dirkhh commented:
With the latest push I have the keyboard opening reliably, but sometimes (on Android) the text input field isn't shown until I scroll "up" a tiny bit in the dive list (i.e., drag my finger down to show newer dives). And there is an ugly UI glitch where it stays frozen for a bit with the keyboard removed but the dive list not repainted while it waits to match the right dives.
It should be easy to have it show "something else" there, but as usual with QML, my first couple of attempts didn't work :-)

dirkhh commented:
OK, lots of updates. This might be ready to be merged. It's far from perfect, but I think it's viable.
Known issues:
- on Android I usually need to pull down the dive list a tiny bit before the input field for the filter (the ListView header) is shown. No idea why - I tried MANY things to fix that, no luck).
- the trip headers still show the unfiltered number of dives - I think it may be better with the filtered number, but that's sadly compute intensive to implement.
- sometimes on Android I don't get the virtual keyboard. I tried to hack around that, but it's still not 100%

Please test and comment!

jbygdell commented on mobile-widgets/qml/DiveList.qml:
```diff
@@ -421,6 +421,16 @@ Kirigami.ScrollablePage {
 							forceActiveFocus()
 						}
 					}
+					onFocusChanged: {
```
This block gives another issue, the keyboard is not hidden when pressing to hide the filter view.
Removing the onFocusChanged block solves this issue for me.

dirkhh replied:
That's wild since I actually ask to hide it RIGHT THERE.
But without this block I could easily reproduce a situation where I had no keyboard and tapping on the input field didn't get me a keyboard, either. So I'm not sure what to do here... maybe don't try to explicitly hide it?

dirkhh replied:
Yeah, that's what bugs me the most. How can we write an app that works for all (or at least most) of our users???
Not sure what to do here - is there another way to ensure that the keyboard is shown when the TextField is entered?

jbygdell replied:
Activating the textfield by tapping on it will always bring up the keyboard in my tests. And the forceActiveFocus fixes the issue with displaying the keyboard when opening the filter view.

dirkhh commented:
OK, now performance is MUCH better. That's more like it.
All we have to figure out is that pesky keyboard / focus issue. Oops, forgot to push that patch.

dirkhh commented:
I think this is ready to move to master. @jbygdell @bstoeger @janmulder any concerns?

bstoeger commented on mobile-widgets/qmlmanager.cpp:
```diff
@@ -1847,6 +1849,17 @@ void QMLManager::showDownloadPage(QString deviceString)
+{
+	// show that we are doing something, then do something in another thread in order not to block the UI
+	QMetaObject::invokeMethod(qmlWindow, "showBusy");
+	QFuture<void> future = QtConcurrent::run(QThreadPool::globalInstance(),
```
`future` is never used - am I missing something?

dirkhh replied:
Good catch. I thought I'd use a FutureWatcher to know when this is done - but then ended up simply doing the call to hide in the lambda

(BTW: I can't believe that no one has called me out of using a lambda in my implementation - I was certain that I'd get some serious ribbing over this one...)

bstoeger commented on qt-models/divelistmodel.h:
```diff
@@ -12,11 +12,21 @@ class DiveListSortModel : public QSortFilterProxyModel
+protected:
+	bool filterAcceptsRow(int source_row, const QModelIndex &source_parent) const;
+private:
+	std::vector<bool> filteredRows;
```
Not important, but note that `std::vector<bool>` has a design mistake which turns it into a bitfield (!). Thus, it tends to be much less efficient than one might expect.

dirkhh replied:
Oh, so I should make this a std::vector<char> or something like that?

dirkhh replied:
The reason I'm asking, in case that's not obvious... the main reason to "roll my own" implementation of the filter was performance, so I really wouldn't want to to then use a poor implementation of a vector and lose some of the advantage I'm trying to get...

bstoeger replied:
> Oh, so I should make this a std::vector or something like that?

Out of principle, probably yes. Realistically, the microsecond more-or-less will not be perceivable.

bstoeger commented:
At a quick glance it looks reasonable, but I'm no well-versed enough in QML to form a proper opinion.

dirkhh commented:
@janmulder thanks for the review and encouragement

at 1 & 2) you don't think it's ADORABLE? Ok, maybe not. Yeah, that thing is obviously entirely unstyled and there is a ton of room for improvement. No argument.
at 3) that's odd - it should have height zero when not in use. There was some visual oddity when I made it invisible early in my tests, but again, easy enough to play with some more.
at 4) yes. I actually looked into fixing that and that gave me the idea for the much faster filter... and then I forgot to fix those numbers.
at 4) (you had two 4s) I hadn't noticed that it wasn't reset - should also be easy to fix. I'd love to hear other design ideas how that result should be presented (but I think it's an important one because that way you can easily answer questions like "how many dives did I do in my dry suit")

dirkhh commented:
manually merged. I couldn't figure out how to change the color of the busy indicator without completely rolling my own. What the heck? But at least I made it smaller.
And I fixed the bug with the count of found elements not being reset.
So (1) and the the first half of the second (4) above are addressed. The rest still need to be done. But I merged it into master for now. I'll be spending most of tomorrow on airplanes, so feel free to play with this.
